lib/scholar/decomposition/pca.ex

```diff
@@ -0,0 +1,216 @@
+defmodule Scholar.Decomposition.PCA do
+  @moduledoc """
+  PCA decomposition algorithm.
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container,
+           containers: [
+             :components,
+             :explained_variance,
+             :explained_variance_ratio,
+             :singular_values,
+             :mean,
+             :num_components,
+             :num_features,
+             :num_samples,
+             :decomposer
+           ]}
+  defstruct [
+    :components,
+    :explained_variance,
+    :explained_variance_ratio,
+    :singular_values,
+    :mean,
+    :num_components,
+    :num_features,
+    :num_samples,
+    :decomposer
+  ]
+
+  fit_opts_schema = [
+    num_components: [
+      type:
+        {:or, [{:custom, Scholar.Options, :positive_number, []}, {:in, [:none, :pos_integer]}]},
+      default: :none,
+      doc: ~S"""
+      Number of components to keep. if `:num_components` is not set all components are kept:
+
+      $num\\_components = min(num\\_samples, num\\_features)$
+
+      If $0 < num\\_components <= 1$, select the number of components such
+      that the amount of variance that needs to be explained is greater than the percentage specified by `:num_components`.
```

Comment:
What does the `:pos_integer` value does?

Reply:
If `:num_components` is an integer then it simply indicates the number of components to keep.